### PR TITLE
Refactor file-browser component and use in select pages

### LINKED ISSUE #4347: Transmission Downloaded Files & PP-Seeding Paths Getting Corrupted
Medusa Info: | Branch: master
Commit: c65ca2607fae74405dcceaaf8c02f3d29d0313ea                     Version: v0.2.5                     Database: 44.11
Python Version: | 2.7.14 (v2.7.14:84471935ed, Sep 16 2017, 20:25:58) [MSC v.1500 64 bit (AMD64)]
SSL Version: | OpenSSL 1.0.2k  26 Jan 2017
OS: | Windows-7-6.1.7601-SP1
Locale: | en_GB.cp1252

Just noticed this issue last night.  And, Medusa isn't registering downloads when snatched torrents are downloaded.

**What you did:**
Enter: Search Settings>Torrent Search>Downloaded files location and Post-Processed seeding torrents location = `D\:ADMIN\comp\tv`.
**What happened:**
Save and refresh page and get literally `D:ADMINcomp	v`

(config/postProcessing/ PP Dir is maintaining correct path)
**What you expected:**
Really?
**Logs:**
```
PASTE LOGS HERE
No errors show in log

2018-06-12 07:24:03 DEBUG    Thread_24 :: [c65ca26] Broken providers already updated in the last hour
2018-06-12 07:24:01 DEBUG    Thread_23 :: [c65ca26] Broken providers already updated in the last hour
2018-06-12 07:23:13 INFO     POSTPROCESSOR :: [c65ca26] Successfully processed.
2018-06-12 07:23:13 DEBUG    POSTPROCESSOR :: [c65ca26] No processable items found in folder: D:\ADMIN\comp\tv
2018-06-12 07:23:13 DEBUG    POSTPROCESSOR :: [c65ca26] Processing path: D:\ADMIN\comp\tv
2018-06-12 07:23:13 DEBUG    POSTPROCESSOR :: [c65ca26] Starting new thread: POSTPROCESSOR
2018-06-12 07:18:49 DEBUG    Thread_9 :: [c65ca26] Broken providers already updated in the last hour
2018-06-12 07:18:41 DEBUG    Thread_4 :: [c65ca26] Broken providers already updated in the last hour
2018-06-12 07:18:13 DEBUG    TRAKTCHECKER :: [c65ca26] Starting new thread: TRAKTCHECKER
2018-06-12 07:18:04 DEBUG    Thread_3 :: [c65ca26] Broken providers already updated in the last hour
2018-06-12 07:17:58 DEBUG    Thread_6 :: [c65ca26] Broken providers already updated in the last hour
2018-06-12 07:17:31 DEBUG    Thread_4 :: [c65ca26] Broken providers already updated in the last hour
2018-06-12 07:17:26 DEBUG    Thread_0 :: [c65ca26] Broken providers already updated in the last hour
2018-06-12 07:15:36 DEBUG    Thread_2 :: [c65ca26] Broken providers already updated in the last hour
2018-06-12 07:15:27 DEBUG    Thread_0 :: [c65ca26] Broken providers already updated in the last hour
2018-06-12 07:13:34 DEBUG    Thread_0 :: [c65ca26] Broken providers already updated in the last hour
2018-06-12 07:13:20 DEBUG    CHECKVERSION :: [c65ca26] User-Agent: Medusa/0.2.5 (Windows; 7; 4b0cac91-6dff-11e8-90a1-005056c00008)
2018-06-12 07:13:20 DEBUG    CHECKVERSION :: [c65ca26] GET URL: https://cdn.pymedusa.com/news/news.md [Status: 200]
2018-06-12 07:13:20 DEBUG    CHECKVERSION :: [c65ca26] Checking GitHub for latest news.
2018-06-12 07:13:20 DEBUG    CHECKVERSION :: [c65ca26] cur_commit = c65ca2607fae74405dcceaaf8c02f3d29d0313ea, newest_commit = c65ca2607fae74405dcceaaf8c02f3d29d0313ea, num_commits_behind = 0, num_commits_ahead = 0
2018-06-12 07:13:20 DEBUG    CHECKVERSION :: [c65ca26] git rev-list --left-right "@{upstream}"...HEAD : returned successful
2018-06-12 07:13:19 DEBUG    CHECKVERSION :: [c65ca26] Executing git rev-list --left-right "@{upstream}"...HEAD with your shell in C:\Medusa
2018-06-12 07:13:19 DEBUG    CHECKVERSION :: [c65ca26] git rev-parse --verify --quiet "@{upstream}" : returned successful
2018-06-12 07:13:19 DEBUG    CHECKVERSION :: [c65ca26] Executing git rev-parse --verify --quiet "@{upstream}" with your shell in C:\Medusa
2018-06-12 07:13:19 DEBUG    CHECKVERSION :: [c65ca26] git fetch --prune origin : returned successful
2018-06-12 07:13:17 DEBUG    CHECKVERSION :: [c65ca26] Executing git fetch --prune origin with your shell in C:\Medusa
2018-06-12 07:13:17 DEBUG    CHECKVERSION :: [c65ca26] git config remote.origin.pushurl https://github.com/pymedusa/Medusa : returned successful
2018-06-12 07:13:17 DEBUG    CHECKVERSION :: [c65ca26] Executing git config remote.origin.pushurl https://github.com/pymedusa/Medusa with your shell in C:\Medusa
2018-06-12 07:13:17 DEBUG    CHECKVERSION :: [c65ca26] git config remote.origin.url https://github.com/pymedusa/Medusa : returned successful
2018-06-12 07:13:17 DEBUG    CHECKVERSION :: [c65ca26] Executing git config remote.origin.url https://github.com/pymedusa/Medusa with your shell in C:\Medusa
2018-06-12 07:13:17 DEBUG    CHECKVERSION :: [c65ca26] git rev-parse HEAD : returned successful
2018-06-12 07:13:17 DEBUG    CHECKVERSION :: [c65ca26] Executing git rev-parse HEAD with your shell in C:\Medusa
2018-06-12 07:13:17 DEBUG    CHECKVERSION :: [c65ca26] git symbolic-ref -q HEAD : returned successful
2018-06-12 07:13:17 DEBUG    CHECKVERSION :: [c65ca26] Executing git symbolic-ref -q HEAD with your shell in C:\Medusa
2018-06-12 07:13:17 DEBUG    CHECKVERSION :: [c65ca26] git ls-remote --heads origin : returned successful
2018-06-12 07:13:15 DEBUG    MAIN :: [c65ca26] Finished updating network timezones
2018-06-12 07:13:15 DEBUG    MAIN :: [c65ca26] User-Agent: Medusa/0.2.5 (Windows; 7; 4b0cac91-6dff-11e8-90a1-005056c00008)
2018-06-12 07:13:15 DEBUG    MAIN :: [c65ca26] GET URL: https://cdn.pymedusa.com/sb_network_timezones/network_timezones.txt [Status: 200]
2018-06-12 07:13:15 DEBUG    MAIN :: [c65ca26] Started updating network timezones

### LINKED ISSUE #4355: http://localhost:8081/config/search/ Test SABnzbd Button is dead
Test NZBget and Black hole test ok, but Test SABnzbd Button is dead is dead.  Logs don't show an error. Nor does dev console view in Firefox 60.0.2 (64-bit).  Also, checked config.ini for SABnzb info...  Host and api key don't update the values in the file.  Related to windows file path issue or maybe I have a corrupted db?

### PR DESCRIPTION
**Pull request by p0psicles**
- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read the [contribution guide](https://github.com/pymedusa/Medusa/blob/master/.github/CONTRIBUTING.md)

:heavy_check_mark: ~Should be rebased on develop after merge of #4055~

- Refactors file-browser to handle multiple browsers on the the same page,
  and to rely less on jQuery and more on Vue.
- Use file-browser in `config_search.mako` and `config_backuprestore.mako`.
- Fix #4347 - due to faulty Python to JS conversion. Always use `json.dumps`...
- Fix #4355 and more issues in `config_search.mako`.

### CODE REVIEW

File: themes-default/slim/views/config_backuprestore.mako [hunk 1]
```diff
@@ -97,7 +81,8 @@ const startVue = () => {
                         <div class="field-pair">
                             Select the folder you wish to save your backup file to:
                             <br><br>
-                            <input type="text" name="backupDir" id="backupDir" class="form-control input-sm input350"/>
+                            <!-- <input v-model="backup.dir" type="text" name="backupDir" id="backupDir" class="form-control input-sm input350"/>-->
+                            <file-browser name="backupDir" ref="backupDirBrowser" id="backupDir" title="Select Show Location" :initial-dir="backup.dir" @update:location="backup.dir = $event"></file-browser>
```
Comment (sharkykh) [inline themes-default/slim/views/config_backuprestore.mako]: 1. I don't remember why I added the `ref=` in editShow but it's not needed.
`id=` is not needed here either.
2. `initial-dir` is not needed in this case, as it will always be empty (no config variable.
3. You need to add `local-storage-key="backupDir"`

File: themes-default/slim/views/config_backuprestore.mako [hunk 2]
```diff
@@ -113,7 +98,8 @@ const startVue = () => {
                         <div class="field-pair">
                             Select the backup file you wish to restore:
                             <br><br>
-                            <input type="text" name="backupFile" id="backupFile" class="form-control input-sm input350"/>
+                            <!-- <input v-model="restore.file" type="text" name="backupFile" id="backupFile" class="form-control input-sm input350"/> -->
+                            <file-browser name="backupFile" ref="backupFileBrowser" id="backupFile" title="Select Show Location" :initial-dir="restore.file" :include-files="true" @update:location="restore.file = $event"></file-browser>
```
Comment (sharkykh) [inline themes-default/slim/views/config_backuprestore.mako]: 1. I don't remember why I added the `ref=` in editShow but it's not needed.
`id=` is not needed here either.
2. `initial-dir` is not needed in this case, as it will always be empty (no config variable.
3. You need to add `local-storage-key="backupFile"`
4. As explained in the toggle button PR, boolean properties can be `<file-browser ... include-files>`.

File: themes-default/slim/views/config_backuprestore.mako (hunk 1 above)
Comment (p0psicles) [reply themes-default/slim/views/config_backuprestore.mako]: 1 and 2 i get. But why do i need local storage key here?

File: themes-default/slim/views/config_backuprestore.mako (hunk 2 above)
Comment (p0psicles) [reply themes-default/slim/views/config_backuprestore.mako]: 4: didnt know tnx

File: themes-default/slim/views/vue-components/file-browser.mako [hunk 3]
```diff
@@ -138,8 +140,8 @@ Vue.component('file-browser', {
     },
     mounted() {
         const vm = this;
-
         let fileBrowserDialog;
+        this.currentPath = this.initialDir;
```
[the remark this answers is not in the document]
Comment (p0psicles) [reply themes-default/slim/views/vue-components/file-browser.mako]: This didnt work here. Thats whay i added. I dont know why the merhod was not triggered for this page

File: themes-default/slim/views/config_search.mako [hunk 4]
```diff
@@ -582,7 +578,7 @@ const startVue = () => {
                                     <label>
                                         <span class="component-title">Black hole folder location</span>
                                         <span class="component-desc">
-                                            <input type="text" name="nzb_dir" id="nzb_dir" v-model="nzb.dir" class="form-control input-sm input350"/>
+                                            <file-browser name="nzb_dir" id="nzb_dir" title="Select nzb black hole Location" :initial-dir="nzb.dir" @update:location="nzb.dir = $event"></file-browser>
```
[the remark this answers is not in the document]
Comment (p0psicles) [reply themes-default/slim/views/config_search.mako]: Sure

File: themes-default/slim/views/vue-components/file-browser.mako (hunk 3 above)
[the remark this answers is not in the document]
Comment (sharkykh) [reply themes-default/slim/views/vue-components/file-browser.mako]: Which page is that?

[the remark this answers is not in the document]
Comment (p0psicles) [reply themes-default/slim/views/vue-components/file-browser.mako]: SearchResults

[the remark this answers is not in the document]
Comment (sharkykh) [reply themes-default/slim/views/vue-components/file-browser.mako]: Oh I think I know why!
I added that watch to make sure if initialDir is updated AFTER the component is created (if you get the value from an API call) it would update the current value.
In this case the value isn't updated because it's available directly from Python.

**TL;DR:** You're correct in adding this line, but I think changing `currentPath` in `data()` to `this.initialDir` is better :) (On line 84)
